### PR TITLE
Fix BCOO gradients for vmapped sparse matvecs

### LINKED ISSUE #37647: `jax.experimental.sparse.bcoo._bcoo_dot_general_transpose` returns a broadcasted cotangent for vmapped sparse matvecs
### Description

Reverse-mode differentiation of vmapped `BCOO` matrix-vector products fails because the transpose rule for `bcoo_dot_general` returns a cotangent for `lhs_data` with the fully broadcast batch shape, instead of reducing it back to the primal sparse-data shape.

#### Minimal Reproducer

```python
import jax
import jax.numpy as jnp
from jax.experimental.sparse import BCOO


A_dense = jnp.array(
    [
        [4.0, -1.0, 0.0, 0.0],
        [-1.0, 4.0, -1.0, 0.0],
        [0.0, -1.0, 4.0, -1.0],
        [0.0, 0.0, -1.0, 4.0],
    ],
    dtype=jnp.float32,
)
A_bcoo = BCOO.fromdense(A_dense)
B = jnp.array(
    [
        [1.0, 0.0, 2.0],
        [0.0, 1.0, 1.0],
        [1.0, 1.0, 0.0],
        [0.0, 2.0, 1.0],
    ],
    dtype=jnp.float32,
)


def objective(data):
    A = BCOO(
        (data, A_bcoo.indices),
        shape=A_bcoo.shape,
        indices_sorted=A_bcoo.indices_sorted,
        unique_indices=A_bcoo.unique_indices,
    )
    X = jax.vmap(lambda b: A @ b, in_axes=1, out_axes=1)(B)
    return jnp.sum(X**2)


print(jax.grad(objective)(A_bcoo.data))
```

The failure is specific to the vmapped sparse matvec reverse path.

- `grad(sum(vmap(lambda b: A_bcoo @ b)(B)**2))` fails
- `grad(sum((A_bcoo @ B)**2))` succeeds
- dense `A_dense @ b` under the same `vmap` pattern succeeds

So the issue is not generic `vmap` reverse mode, and not generic callback batching. It is specific to the `BCOO` vmapped matvec transpose path. As reported below, `jax` version < 0.9 do not error, so the bug has been introduced in > 0.9 versions.

### System info (python version, jaxlib version, accelerator, etc.)

- `jax==0.9.0`, `0.9.1`, `0.9.2`, `0.10.0`: reproduced
- `jax==0.8.0`: did not reproduce

### PR DESCRIPTION
**Pull request by lanarkite99**
Fixes #37647

This PR fixes reverse-mode differentiation through vmapped `BCOO` matrix-vector products.

The transpose rule for `bcoo_dot_general` can produce a cotangent for `lhs_data` with broadcast batch dimensions. The primal sparse data may have size-1 batch dimensions, so returning the broadcasted cotangent directly leads to a shape mismatch during transpose checking.

This change reuses JAX's existing `_unbroadcast` helper before returning the sparse-data cotangent, reducing it back to the primal `lhs_data` shape.

### CODE REVIEW

**Review by jakevdp — APPROVED** [review]
Thanks, looks great!
